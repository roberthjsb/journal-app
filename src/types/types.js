
export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError : '[UI] Set Error',
    uiRemoveError : '[UI] Remove Error',
    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',

    notesAddNew: "[Notes] New note",
    notesActive: "[Notes] Set Active note",
    noteLoad: '[Notes] Load notes',
    notesUpdated: '[Notes] Updated note note',
    notesFileUrl: '[Notes] Updated image url',
    noteDelete: '[Notes] Delete note',
    noteLogoutCleaning: '[Notes] Logout Cleaning',

};