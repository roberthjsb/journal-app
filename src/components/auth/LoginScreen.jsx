import React from "react";
import { Link } from "react-router-dom";
import { useForm } from "./../../hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import {
  startLoginEmailPassword,
  startGoogleLogin,
} from "./../../actions/auth";

export const LoginScreen = () => {
  
  const {loading} = useSelector(state=> state.ui);
  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    email: "roberth@gmail.com",
    password: "123456",
  });
  const { email, password } = formValues;
  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };
  const handleGoogleLogin = () => {
    loading || dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className="auth__title">Login</h3>
      <form onSubmit={handleLogin}>
        <input
          className="auth__input"
          type="text"
          name="email"
          placeholder="Email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          name="password"
          placeholder="password"
          value={password}
          onChange={handleInputChange}
        />
        <button 
          className="btn btn-primary btn-block" 
          type="submit" 
          disabled={ loading }
        >
          Login
          { 
          loading &&
          <div class="lds-ripple"><div></div><div></div></div>
          }

        </button>
        <div className="auth__social-networks">
          <p>Login with Social Networks</p>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link to="/auth/register" className="link">
          Create new account
        </Link>
      </form>
    </>
  );
};
