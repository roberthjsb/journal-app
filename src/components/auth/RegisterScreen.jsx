import React from 'react'
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
    return (
        <>
        <h3 className="auth__title">Login</h3>
        <form action="">
        <input
            className="auth__input"
            type="text"
            name="nombre"
            placeholder="Nombre"
            autoComplete="off"
          />
          <input
            className="auth__input"
            type="text"
            name="email"
            placeholder="Email"
            autoComplete="off"
          />
          <input
            className="auth__input"
            type="password"
            name="password"
            placeholder="password"
          />
          <input
            className="auth__input"
            type="password"
            name="confirmacion"
            placeholder="ConfirmaciÃ³n"
          />
          <button 
          className="btn btn-primary btn-block mb-5"
            type="submit">Register</button>          
          <Link 
            to="/auth/login"
            className="link"
          >
              Already registered?
          </Link>
        </form>
      </>
    )
}
