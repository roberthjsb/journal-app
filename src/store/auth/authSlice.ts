import { createSlice, PayloadAction } from "@reduxjs/toolkit"

type authState = {
    status: 'checking' | 'not-authenticated' | 'authenticated',
    uid: string | null,
    email: string | null,
    displayName: string | null,
    photoURL: string | null,
    errorMessage: string | null
}
const initialState: authState = {
    status: 'not-authenticated',
    uid: null,
    email: null,
    displayName: null,
    photoURL: null,
    errorMessage: null
}

export const authSlice = createSlice({
    name: 'auth',
    initialState,
    reducers: {
        login: (state, action: PayloadAction<authState>) => {
            state.status = 'authenticated';
            state.uid = action.payload.uid;
            state.email = action.payload.email;
            state.displayName = action.payload.displayName;
            state.photoURL = action.payload.photoURL;
            state.errorMessage = null
        },
        logout: (state, action: PayloadAction<authState>) => {
            state.status = 'not-authenticated';
            state.uid = null;
            state.email = null;
            state.displayName = null;
            state.photoURL = null;
            state.errorMessage = action.payload.errorMessage
        },
        checkingCredencials: (state) => {
            state.status = 'checking'
        }
    }
})

export const { login, logout, checkingCredencials } = authSlice.actions