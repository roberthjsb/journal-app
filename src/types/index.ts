type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;


export type JournalNote = {
    id: string,
    title: string,
    body: string,
    date?: number,
    imageUrls: string[]
}

export type Journal = {
    isSaving: boolean,
    messageSaved: string,
    notes: JournalNote[],
    active: JournalNote | null
}


export type statusState={ status?: 'checking' | 'not-authenticated' | 'authenticated'}
export type errorState= { errorMessage?: string | null}
export type AuthUserState = {
    uid?: string | null,
    email?: string | null,
    displayName?: string | null,
    photoURL?: string | null,
}
export type AuthState= statusState & AuthUserState & errorState;
export type StatusResponse = {ok:boolean};
export type AuthResult = AuthState & StatusResponse

export type StoreState = {
    auth:AuthState,
    journal:Journal
}
