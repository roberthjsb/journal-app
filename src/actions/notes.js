import { db } from "../firebase/firebaseConfig";
import { types } from './../types/types';
import { loadNotes } from './../helper/loadNotes';
import Swal from 'sweetalert2'
import { fileUpload } from './../helper/fileUpload';

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        const doc = await db.collection(`${uid}/journal/notes`)
            .add(newNote);
        dispatch(activeNote(doc.id, newNote));
        dispatch(addNote(doc.id, newNote));

    }
}
export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const addNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLooadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}


export const setNotes = (notes) => ({
    type: types.noteLoad,
    payload: notes
});

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        debugger
        if (!note.url) {
            delete note.url;
        }
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
        dispatch(refreshNote(note.id, noteToFirestore))
        Swal.fire('Saved', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active } = getState().notes;
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            onBeforeOpen:()=>{
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        active.url= fileUrl;
        dispatch(startSaveNote(active))
        Swal.close();

    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        await db.doc(`${uid}/journal/notes/${id}`).delete();
        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => {
    return {
        type: types.noteDelete,
        payload: id
    }
}
export const notLogout = () => {
    return {
        type: types.noteLogoutCleaning
    }
}




