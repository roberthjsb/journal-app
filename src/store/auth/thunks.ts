import { loginWithEmailAndPassword, logoutSession, registerUserWithCredential, signInWithGoogle } from "../../firebase/providers";
import { AuthState } from "../../types";
import { clearJournal } from "../journal/journalSlice";
import { AppDispatch } from "../store";
import {  checkingCredencials, login, logout } from "./authSlice";

export const checkingAuthentication = () => {
    return async (dispatch: AppDispatch) => {
        dispatch(checkingCredencials());
    }
}

export const startGoogleSignIn = () => {
    return async (dispatch: AppDispatch) => {
        dispatch(checkingCredencials());

        const result = await signInWithGoogle();

        if (!result.ok) return dispatch(logout(result as AuthState))
        dispatch(login(result))
    }
}

export const startRegisterUserWithCredential = (credencial: { email: string, password: string, displayName: string }) => {
    return async (dispatch: AppDispatch) => {
        dispatch(checkingCredencials());
        const { ok, ...authInformation } = await registerUserWithCredential(credencial)
        if (!ok) return dispatch(logout(authInformation))
        dispatch(login(credencial))
    }
}


export const startLoginUserWithEmailAndPassword = (credencial: { email: string, password: string }) => {
    return async (dispatch: AppDispatch) => {

        dispatch(checkingCredencials());
        const { ok, ...authInformation } = await loginWithEmailAndPassword(credencial);
        if (!ok) return dispatch(logout(authInformation as AuthState))
        dispatch(login(authInformation))
    }
}

export const StartLogut = () => {
    return async (dispatch: AppDispatch) => {
        await logoutSession();
        dispatch(clearJournal())
        dispatch(logout());
    }
}



