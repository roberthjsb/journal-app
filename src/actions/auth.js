import { types } from './../types/types';
import { firebase, googleAuthProvider } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import Swal from 'sweetalert2';
import { notLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading())
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName))                
            })
            .catch(e => {
                Swal.fire('Error',e.message,'error')
            })
            .finally(()=>dispatch(finishLoading()))
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName))
            }).catch(e=>console.log(e))
    }
}

export const registerWithEmailAndPassword = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password.toString())
            .then(async ({ user }) => {
                console.log('createUserWithEmailAndPassword')
                await user.updateProfile({ displayName: name });
                dispatch(login(user.uid, user.displayName));
            }).catch(e=>console.log(e))
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: { uid, displayName }
})
export const startLogout = () => {
    console.log('startLogout')
    return async  (dispatch)=>{
        console.log('startLogout dispatch')
        await firebase.auth().signOut()
        dispatch(logout())
        dispatch(notLogout())
    }
}

export const logout = () => ({
    type: types.logout,
})
