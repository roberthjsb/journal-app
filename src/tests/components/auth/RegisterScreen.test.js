import React from 'react'
import { mount } from 'enzyme';
import '../../../setupTests';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { MemoryRouter } from 'react-router-dom'
import { RegisterScreen } from './../../../components/auth/RegisterScreen';
import { types } from './../../../types/types';
import { registerWithEmailAndPassword } from './../../../actions/auth';

const middlewares = [thunk];
let mockStore;
let store;
let wrapper;
const initialStore = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};
jest.mock('../../../actions/auth', () => ({
    registerWithEmailAndPassword: jest.fn()
}))


describe('RegisterScreen test', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        mockStore = configureStore(middlewares);
        store = mockStore(initialStore);
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>);
    })

    it('should be display RegisterScreen', () => {
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('h3.auth__title')).toBeTruthy()
    });
    it('should be change state de msgError for email invalid', () => {
        const inputField = wrapper.find('input[name="email"]');
        inputField.simulate('change', {
            target: {
                name: 'email',
                value: ''
            }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault: () => { }
        });
        const actions = store.getActions();
        //Email is not valid
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        });
    });
    it('should be change state de msgError for name ', () => {
        const inputField = wrapper.find('input[name="name"]');
        inputField.simulate('change', {
            target: {
                name: 'name',
                value: ''
            }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault: () => { }
        });
        const actions = store.getActions();
        //Email is not valid
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Name is require'
        });
    });

    it('should be change state de msgError for password validation ', () => {
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>);
        const inputField = wrapper.find('input[name="password"]');
        inputField.simulate('change', {
            target: {
                name: 'password',
                value: ''
            }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault: () => { }
        });
        expect(store.getActions()[0]).toEqual({
            type: types.uiSetError,
            payload: 'Password should ve at least 6 characters and match password with confirmation'
        })
    });

    it('should be Display error message in state ', () => {
        const initialStore = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Error email invalido'
            }
        };
        store = mockStore(initialStore);
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>);
        expect(wrapper.find('.auth__alert-error').text()).toBe(initialStore.ui.msgError)
    })

    it('should be remove error message in state ', () => {
        const initialStore = {
            auth: {},
            ui: {
                loading: false,
                msgError: null
            }
        };
        
        store = mockStore(initialStore);
        store.dispatch= jest.fn();
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>);

        wrapper.find('form').simulate('submit', {
            preventDefault: () => { }
        });
        expect(registerWithEmailAndPassword).toBeCalledWith("roberth.salazar@outlook.com",123456,"Roberth J");
    })

});