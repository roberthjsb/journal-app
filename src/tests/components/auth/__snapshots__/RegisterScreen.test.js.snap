// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterScreen test should be display RegisterScreen 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Login
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Nombre"
      type="text"
      value="Roberth J"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="roberth.salazar@outlook.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value={123456}
    />
    <input
      className="auth__input"
      name="confirmacion"
      onChange={[Function]}
      placeholder="ConfirmaciÃ³n"
      type="password"
      value={123456}
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
